File: quicksort.py
def quicksort(array):
    return quicksortHelper(array, 0, len(array) - 1)

def quicksortHelper(array, left, right):
    if left >= right:
        return

    pivot = partition(array, left, right)

    quicksortHelper(array, pivot + 1, right)
    quicksortHelper(array, left, pivot - 1)

    return array

def partition(array, left, right):
    pivot_index = left
    pivot = array[pivot_index]
    left += 1

    while left <= right:
        while (array[left] < pivot):
            left += 1

        while (array[right] > pivot):
            right -= 1
        
        if left <= right:
            array[left], array[right] = array[right], array[left]
            left += 1
            right -= 1
    
    array[pivot_index], array[right] = array[right], array[pivot_index]
    return right

print quicksort([5,8,2,4,2,9])

File: quicksort.js
function quickSort(a, l, r) {

  if (l >= r) {
    return a;
  }

  var x = a[l]; // partition at left index
  var i = l;  // i === low index
  var j = r; // j = high index

  while (i <= j) {  // low index <= high index
    while (a[i] < x) {  // while element @ low < partition
      i++; // increase low index
    }
    while (a[j] > x) {  // while the el @ high > x
      j--;  // increase high
    }
    if (i <= j) {  // if low index < high
      var t = a[i];  //swap
      a[i] = a[j];
      a[j] = t;
      i++;  // increase low
      j--;  // decrease high
    }
  }

  quickSort(a, l, i - 1);
  quickSort(a, i, r);

  return a;
}


const result = quickSort([5, 2, 1, 7, 5, 3, 2, 3], 0, 7);
console.log(result);
